From: Felix Fietkau <nbd@nbd.name>
Date: Tue, 22 Jul 2025 11:58:15 +0200
Subject: [PATCH] gcc: prevent the use of LDRD/STRD on ARMv5TE

These instructions are for 64-bit load/store. On ARMv5TE, the CPU
requires addresses to be aligned to 64-bit. When misaligned, behavior is
undefined (effectively either loads the same word twice on LDRD, or
corrupts surrounding memory on STRD).

On ARMv6 and newer, unaligned access is safe.

Removing these instructions for ARMv5TE is necessary, because GCC
ignores alignment information in pointers and does unsafe optimizations
that have shown up as bugs in various places.

This patch was originally added more than 11 years ago in commit b050f87d13b5,
but got lost 6 years ago, when gcc 9.1 was added in 88c07c655262.

This primarily affects the kirkwood and ixp4xx targets

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---
--- a/gcc/config/arm/arm.h
+++ b/gcc/config/arm/arm.h
@@ -165,7 +165,7 @@ emission of floating point pcs attribute
 /* Thumb-1 only.  */
 #define TARGET_THUMB1_ONLY		(TARGET_THUMB1 && !arm_arch_notm)
 
-#define TARGET_LDRD			(arm_arch5te && ARM_DOUBLEWORD_ALIGN \
+#define TARGET_LDRD			(arm_arch6 && ARM_DOUBLEWORD_ALIGN \
                                          && !TARGET_THUMB1)
 
 #define TARGET_CRC32			(arm_arch_crc)
